```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=TRUE)
```
# Methylation Analysis with Epiviz Integration

## Introduction:

#### Purpose: 
To analyze Illumina 450k Methylation Array data to find differences in methylation at the TRIM29 gene located on chromosome 11 between various tumor samples (breast, lung, colon, endometrium) and between normal and tumor breast cancer samples. 

#### How:  
Using the minfi and epivizr packages in Bioconductor and Epiviz as an interactive genomic tool, this protocol aims to walk through the analysis of the Illumina HumanMethylation450 Array data from the Cancer Genome Atlas (TCGA) project.

#### Why:
The TRIM29 gene is of interest because of its opposing functions in various cancers.  In some, it acts as a tumor suppressor (in breast cancers), and in others, it acts as an oncogene (pancreatic and lung cancers).  

#### Workflow:
<<<<<<< HEAD
Packages needed: Bioconductor, minfi, epivizr, GenomicFeatures (within Bioconductor)

```{r,eval=F}
source("http://bioconductor.org/biocLite.R")
biocLite("minfi")
biocLite("epivizr")
biocLite("GenomicFeatures")
biocLite("BSgenome.Hsapiens.UCSC.hg19")
biocLite("TxDb.Hsapiens.UCSC.hg19.knownGene")
```
=======
Packages needed: minfi, epivizr, TCGA-Assembler
>>>>>>> origin/master

* Make a flowchart of overall procedure

download/save raw Illumina 450k Methylation Array data as IDAT files ->  
install necessary packages in R (Bioconductor, minfi, etc.) ->    
preprocess data into final GenomicRatioSet output ->   
find differentially methylated positions using dmpFinder method ->   
find small differentially methylated regions using bumphunter method ->  
find larger differentially methylated regions using blockFinder method

## Procedure:

### 1. Gather data:

<<<<<<< HEAD
In order to analyze the data, it must first be loaded into R.

Download TCGA 450k methylation array using the data matrix in the data portal

https://tcga-data.nci.nih.gov/tcga/dataAccessMatrix.htm

When selecting data, choose samples that end in 01 (TCGA-..-....-01) for solid tissue tumor samples and samples that end in 11 (TCGA-..-....-11) for solid tissue normal samples.

For this protocol, only breast tumor and breast normal tissue (from BRCA), colon normal (from COAD), lung normal (from LUSC) is loaded.

Follow along with the screenshots:


Settings for breast tumor and breast normal download
  
<img src="/Users/mwalter/Documents/screenshots/brca_mat.png" width="600" height="600" />  
  
Chosen breast samples for the analysis 
  
<img src="/Users/mwalter/Documents/screenshots/brca_tum_dat.png" width="600" height="600" />
  
<img src="/Users/mwalter/Documents/screenshots/brca_norm_dat.png" width="600" height="600" /> 
  
Settings for colon normal download
  
<img src="/Users/mwalter/Documents/screenshots/coad_mat.png" width="600" height="600" />
  
Chosen colon samples for the analysis 
  
<img src="/Users/mwalter/Documents/screenshots/coad_norm_dat.png" width="600" height="600" /> 
  
Settings for lung normal download
  
<img src="/Users/mwalter/Documents/screenshots/lusc_mat.png" width="600" height="600" />
  
Chosen lung samples for the analysis 
  
<img src="/Users/mwalter/Documents/screenshots/lusc_norm_dat.png" width="600" height="600" />

For each cancer type, once the correct settings are selected, click the "Build Archive" button and then download the data.

After unzipping the files that are emailed to you, save each cancer type (BRCA, COAD, LUSC) as different folders in "datadir", like shown below.

Load breast tumor and breast normal data

```{r,eval=T}
library(minfi)
datadir <- "/Users/mwalter/Documents/methylation_files/breast"

clinicalDir <- file.path(datadir,"Clinical/Biotab")
sample_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_sample_brca.txt"),sep="\t",stringsAsFactors=FALSE)
keep <- sample_tab$sample_type %in% c("Primary Tumor", "Solid Tissue Normal")
sample_tab <- sample_tab[keep,]

patient_id <- unique(sapply(strsplit(sample_tab$bcr_sample_barcode,split="-"), function(x) paste(x[1:3],collapse="-")))

tumor_sample_id <- sample_tab$bcr_sample_uuid[sample_tab$sample_type=="Primary Tumor"]
normal_sample_id <- sample_tab$bcr_sample_uuid[sample_tab$sample_type=="Solid Tissue Normal"]

# read tumor data
tumor_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_tumor_sample_brca.txt"),sep="\t",stringsAsFactors=FALSE)

tab <- merge(sample_tab, tumor_tab, by="bcr_sample_uuid", suffixes=c(".sample",".tumor"),all.x=TRUE)

# read normal data
normal_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_normal_control_brca.txt"),sep="\t",stringsAsFactors=FALSE)
tab <- merge(tab, normal_tab, by="bcr_sample_uuid", suffixes=c(".tumor",".normal"),all.x=TRUE)

tab$bcr_patient_barcode <- tab$bcr_patient_barcode.tumor

ii <- is.na(tab$bcr_patient_barcode)
tab$bcr_patient_barcode[ii] <- tab$bcr_patient_barcode.normal[ii]

# read patient data
patient_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_clinical_patient_brca.txt"),sep="\t",stringsAsFactors=FALSE)
names(patient_tab) <- paste("patient",names(patient_tab),sep=".")
tmp <- merge(tab,patient_tab,by.x="bcr_patient_barcode",by.y="patient.bcr_patient_barcode",all.x=TRUE,suffixes=c(".sample",".patient"))
tab <- tmp

# read meth metadata
methMetaDir <- file.path(datadir,"METADATA/JHU_USC__HumanMethylation450")
methMeta_tab <- read.delim(file.path(methMetaDir,"jhu-usc.edu_BRCA.HumanMethylation450.1.9.0.sdrf.txt"),sep="\t",stringsAsFactors=FALSE)

sample_barcode <- sapply(strsplit(methMeta_tab$Comment..TCGA.Barcode.,split="-"),function(x) paste(x[1:4],collapse="-"))
m <- match(tab$bcr_sample_barcode,sample_barcode)
tab$Basename <- gsub("_Grn\\.idat","",methMeta_tab$Array.Data.File[m])
tab <- tab[!is.na(tab$Basename),]

basedir <- file.path(datadir,"DNA_Methylation/JHU_USC__HumanMethylation450/Level_1")
tab$Basename <- file.path(basedir,tab$Basename)
keep <- file.exists(paste(tab$Basename,"_Grn.idat",sep=""))
breast_targets <- tab
objs <- grep("tab",ls(),value=TRUE)
rm(list=objs)
objs <- grep("dir",ls(),value=TRUE,ignore=TRUE)
rm(list=objs)

nms <- names(breast_targets)
targets.breast <- breast_targets[nms]

targets.breast$Status <- factor(ifelse(targets.breast$sample_type=="Primary Tumor","cancer","normal"),levels=c("normal","cancer"))
targets.breast$Tissue <- tolower(targets.breast$patient.tumor_tissue_site)
targets.breast$Sex <- targets.breast$patient.gender
```

Load colon normal data

```{r,eval=T}
datadir <- "/Users/mwalter/Documents/methylation_files/colon"

clinicalDir <- file.path(datadir,"Clinical/Biotab")
sample_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_sample_coad.txt"),sep="\t",stringsAsFactors=FALSE)
keep <- sample_tab$sample_type %in% c("Primary Tumor", "Solid Tissue Normal")
sample_tab <- sample_tab[keep,]

patient_id <- unique(sapply(strsplit(sample_tab$bcr_sample_barcode,split="-"), function(x) paste(x[1:3],collapse="-")))

tumor_sample_id <- sample_tab$bcr_sample_uuid[sample_tab$sample_type=="Primary Tumor"]
normal_sample_id <- sample_tab$bcr_sample_uuid[sample_tab$sample_type=="Solid Tissue Normal"]

# read tumor data
tumor_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_tumor_sample_coad.txt"),sep="\t",stringsAsFactors=FALSE)

tab <- merge(sample_tab, tumor_tab, by="bcr_sample_uuid", suffixes=c(".sample",".tumor"),all.x=TRUE)

# read normal data
normal_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_normal_control_coad.txt"),sep="\t",stringsAsFactors=FALSE)
tab <- merge(tab, normal_tab, by="bcr_sample_uuid", suffixes=c(".tumor",".normal"),all.x=TRUE)

tab$bcr_patient_barcode <- tab$bcr_patient_barcode.tumor

ii <- is.na(tab$bcr_patient_barcode)
tab$bcr_patient_barcode[ii] <- tab$bcr_patient_barcode.normal[ii]

# read patient data
patient_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_clinical_patient_coad.txt"),sep="\t",stringsAsFactors=FALSE)
names(patient_tab) <- paste("patient",names(patient_tab),sep=".")
tmp <- merge(tab,patient_tab,by.x="bcr_patient_barcode",by.y="patient.bcr_patient_barcode",all.x=TRUE,suffixes=c(".sample",".patient"))
tab <- tmp

# read meth metadata
methMetaDir <- file.path(datadir,"METADATA/JHU_USC__HumanMethylation450")
methMeta_tab <- read.delim(file.path(methMetaDir,"jhu-usc.edu_COAD.HumanMethylation450.1.9.0.sdrf.txt"),sep="\t",stringsAsFactors=FALSE)

sample_barcode <- sapply(strsplit(methMeta_tab$Comment..TCGA.Barcode.,split="-"),function(x) paste(x[1:4],collapse="-"))
m <- match(tab$bcr_sample_barcode,sample_barcode)
tab$Basename <- gsub("_Grn\\.idat","",methMeta_tab$Array.Data.File[m])
tab <- tab[!is.na(tab$Basename),]

basedir <- file.path(datadir,"DNA_Methylation/JHU_USC__HumanMethylation450/Level_1")
tab$Basename <- file.path(basedir,tab$Basename)
keep <- file.exists(paste(tab$Basename,"_Grn.idat",sep=""))
colon_targets <- tab
objs <- grep("tab",ls(),value=TRUE)
rm(list=objs)
objs <- grep("dir",ls(),value=TRUE,ignore=TRUE)
rm(list=objs)

nms <- names(colon_targets)
targets.colon <- colon_targets[nms]

targets.colon$Status <- factor(ifelse(targets.colon$sample_type=="Primary Tumor","cancer","normal"),levels=c("normal","cancer"))
targets.colon$Tissue <- tolower(targets.colon$patient.tumor_tissue_site)
targets.colon$Sex <- targets.colon$patient.gender
```

Load lung normal data

```{r,eval=T}
datadir <- "/Users/mwalter/Documents/methylation_files/lung"


clinicalDir <- file.path(datadir,"Clinical/Biotab")
sample_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_sample_lusc.txt"),sep="\t",stringsAsFactors=FALSE)
keep <- sample_tab$sample_type %in% c("Primary Tumor", "Solid Tissue Normal")
sample_tab <- sample_tab[keep,]

patient_id <- unique(sapply(strsplit(sample_tab$bcr_sample_barcode,split="-"), function(x) paste(x[1:3],collapse="-")))

tumor_sample_id <- sample_tab$bcr_sample_uuid[sample_tab$sample_type=="Primary Tumor"]
normal_sample_id <- sample_tab$bcr_sample_uuid[sample_tab$sample_type=="Solid Tissue Normal"]

# read tumor data
tumor_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_tumor_sample_lusc.txt"),sep="\t",stringsAsFactors=FALSE)

tab <- merge(sample_tab, tumor_tab, by="bcr_sample_uuid", suffixes=c(".sample",".tumor"),all.x=TRUE)

# read normal data
normal_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_normal_control_lusc.txt"),sep="\t",stringsAsFactors=FALSE)
tab <- merge(tab, normal_tab, by="bcr_sample_uuid", suffixes=c(".tumor",".normal"),all.x=TRUE)

tab$bcr_patient_barcode <- tab$bcr_patient_barcode.tumor

ii <- is.na(tab$bcr_patient_barcode)
tab$bcr_patient_barcode[ii] <- tab$bcr_patient_barcode.normal[ii]

# read patient data
patient_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_clinical_patient_lusc.txt"),sep="\t",stringsAsFactors=FALSE)
names(patient_tab) <- paste("patient",names(patient_tab),sep=".")
tmp <- merge(tab,patient_tab,by.x="bcr_patient_barcode",by.y="patient.bcr_patient_barcode",all.x=TRUE,suffixes=c(".sample",".patient"))
tab <- tmp

# read meth metadata
methMetaDir <- file.path(datadir,"METADATA/JHU_USC__HumanMethylation450")
methMeta_tab <- read.delim(file.path(methMetaDir,"jhu-usc.edu_LUSC.HumanMethylation450.1.7.0.sdrf.txt"),sep="\t",stringsAsFactors=FALSE)

sample_barcode <- sapply(strsplit(methMeta_tab$Comment..TCGA.Barcode.,split="-"),function(x) paste(x[1:4],collapse="-"))
m <- match(tab$bcr_sample_barcode,sample_barcode)
tab$Basename <- gsub("_Grn\\.idat","",methMeta_tab$Array.Data.File[m])
tab <- tab[!is.na(tab$Basename),]

basedir <- file.path(datadir,"DNA_Methylation/JHU_USC__HumanMethylation450/Level_1")
tab$Basename <- file.path(basedir,tab$Basename)
keep <- file.exists(paste(tab$Basename,"_Grn.idat",sep=""))
lung_targets <- tab
objs <- grep("tab",ls(),value=TRUE)
rm(list=objs)
objs <- grep("dir",ls(),value=TRUE,ignore=TRUE)
rm(list=objs)

nms <- names(lung_targets)
targets.lung <- lung_targets[nms]

targets.lung$Status <- factor(ifelse(targets.lung$sample_type=="Primary Tumor","cancer","normal"),levels=c("normal","cancer"))
targets.lung$Tissue <- tolower(targets.lung$patient.tumor_tissue_site)
targets.lung$Sex <- targets.lung$patient.gender

rm(list=ls()[!(ls() %in% c('targets.breast','targets.colon','targets.lung'))])
```

Merge and read data

```{r,eval=T}
merge <- merge(targets.breast,targets.colon,all=TRUE)
targets <- merge(merge,targets.lung,all=TRUE)
targets <- targets[which(file.exists(paste0(targets$Basename,"_Grn.idat"))),]

memory.limit(size=10000)
rg.set <- read.450k(targets$Basename,verbose=TRUE)
pData(rg.set) <- targets

table(targets$Tissue,targets$Status)
=======
In order to analyze the data, it must first be accessible to R.

Download TCGA 450k methylation data and UCSC expression data using TCGA-Assembler package???
	-http://www.compgenome.org/TCGA-Assembler/
	-ending in 01 -> tumor samples
	-ending in 11 -> normal tissue samples
	-load BRCA (breast cancer and normal breast tissue), COAD (colon cancer), LUSC (lung cancer), UCEC (endometrial cancer) data

```{r,eval=F}
library(minfi)
path <- "/Users/mwalter/Documents/methylation_files"
targets.b.t <- read.delim(file.path (path,"jhu-usc.edu_BRCA.HumanMethylation450.1.9.0.sdrf.txt"),as.is=TRUE)
targets.b.t$Type <- "breast"
targets.b.t$Status <- "tumor"

targets.b.n <- read.delim(file.path (path,"jhu-usc.edu_BRCA.HumanMethylation450.1.9.0.sdrf.txt"),as.is=TRUE)
targets.b.n$Type <- "breast"
targets.b.n$Status <- "normal"

targets.l.n <- read.delim(file.path (path,"jhu-usc.edu_LUSC.HumanMethylation450.1.7.0.sdrf.txt"),as.is=TRUE)
targets.l.n$Type <- "lung"
targets.l.n$Status <- "normal"

targets.c.n <- read.delim(file.path (path,"jhu-usc.edu_COAD.HumanMethylation450.1.9.0.sdrf.txt"),as.is=TRUE)
targets.c.n$Type <- "colon"
targets.c.n$Status <- "normal"

path <- "/Users/mwalter/Documents/methylation_files"
targets <- rbind(targets.b.t,targets.l.n,targets.c.n)
targets <- targets[file.exists(file.path(path, targets$Array.Data.File)),]
targets <- targets[grepl("Red", targets$Array.Data.File),]
targets$Basename <- file.path(path, targets$Array.Data.File)
table(targets$Type,targets$Status)
path <- "/Users/mwalter/Documents/TCGA-Assembler/brca1_DirectoryTraverseResult"
targets <- read.delim(file.path(path,"brca1_DirectoryTraverseResult"))
rg.set <- read.450k.exp(path,verbose=TRUE)

targets <- read.table(list.files(path, pattern="sdrf", full.name=TRUE), header=TRUE, sep="\t")
targets <- targets[file.exists(file.path(path, targets$Array.Data.File)),]
targets <- targets[grepl("Red", targets$Array.Data.File),]
targets$Basename <- gsub("_Red.*$", "", file.path(path, targets$Array.Data.File))
rownames(targets) <- basename(targets$Basename)
head(targets)
>>>>>>> origin/master
```

### 2. Preprocessing:

Process data into usable objects for the minfi package.

<<<<<<< HEAD
Need preprocessing to modify data from IDAT files into useable data sets, like an RGChannelSet, MethylSet, GenomicMethylSet, RatioSet, or GenomicRatioSet for the minfi functions to work. -Most of the analysis will be done on the GenomicRatioSet datatype.
=======
Need preprocessing to modify data from IDAT files into useable data sets, like an RGChannelSet, MethylSet, GenomicMethylSet, RatioSet, or GenomicRatioSet for the minfi functions to work. Most of this analysis will be done on the GenomicRatioSet datatype.
>>>>>>> origin/master

Use minfi's variety of preprocess...() functions to convert data sets into usable forms

Make a flow chart of how to preprocess into useful objects: 


Input                  | Processing Function    | Output           | Analytic Use
---------------------- | ---------------------- | ---------------- | --------------------------
<<<<<<< HEAD
Raw data (IDAT files)  | read.450k()            | RGChannelSet     | output of reading data   
=======
Raw data (IDAT files)  | read.450k.exp()        | RGChannelSet     | output of reading data   
>>>>>>> origin/master
RGChannelSet           | preprocessIllumina()   | MethylSet        | dmpFinder method
MethylSet              | mapToGenome()          | GenomicMethylSet | blockFinder method
GenomicMethylSet       | ratioConvert()         | GenomicRatioSet  | bumphunter method

<<<<<<< HEAD
```{r,eval=T}
=======
```{r,eval=F}
>>>>>>> origin/master
meth.set <- preprocessIllumina(rg.set)
gen.meth.set <- mapToGenome(meth.set)
gen.rat.set <- ratioConvert(gen.meth.set,type="Illumina")
```

### 3. Load Epiviz Workspace:

Use Epiviz to vizualize the results found in R session

<<<<<<< HEAD
Start Epiviz in web, starting from a pre-existing workspace (delete all tracks but "Genes" and "CpG Islands")

```{r, eval=T}
require(epivizr)
mgr <- startEpiviz(workspace="WZCPGTd7Duv",tryPorts=TRUE)
#mgr <- startEpiviz(workspace="Usv3U8lGh7v",tryPorts=TRUE)
#mgr$service()
=======
Start Epiviz in web

```{r, eval=F}
require(epivizr)

mgr <- startEpiviz(tryPorts=TRUE)
mgr$service()
>>>>>>> origin/master
```

### 4. DMP Identificaiton:

<<<<<<< HEAD
Overall, compare normal breast tissue to tumor breast tissue to find small differentially methylated gene regions.
=======
Overall, compare normal breast tissue to normal colon tissue to find small differentially methylated gene regions.

#### A. R Analysis:
>>>>>>> origin/master

Use dmpFinder to find differentially methylated positions to find CpG methylation differences that correspond to a particular phenotype (in this case breast normal and breast tumor)

dmpFinder uses an F-test between two groups of different phenotypes to determine which CpG loci are significantly differentially methylated

Look at difference between normal breast and tumor breast tissues

<<<<<<< HEAD

Start with the genes track at the top

Add promoter regions in 

```{r,eval=T}
library(GenomicFeatures)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
promoter.regions <- promoters(TxDb.Hsapiens.UCSC.hg19.knownGene,upstream=2000,downstream=200)
promoters <- mgr$addDevice(promoter.regions, "Promoters")
```

In Epiviz, add a blocks track to show the location of the DMPs

```{r,eval=T}
library(doParallel)
cores <- detectCores()
registerDoParallel(cores)
meth.set.breast <- meth.set[,which(meth.set$Tissue=="breast")]
dmp <- dmpFinder(meth.set.breast,meth.set.breast$Status,type="categorical")
head(dmp)
dmp.gr <- granges(gen.meth.set[rownames(dmp)])
dmps <- mgr$addDevice(dmp.gr, "Differentially Methylated Positions")
```

Add track of p-values underneath to show which DMPs are significant (not significant if p > 0.05):

```{r, eval=T}
dmp.gr$pval <- -log10(dmp$pval)
pvals <- mgr$addDevice(dmp.gr, "P-Values (-log10) of DMPs",type="bp",columns="pval")
```

Add beta-values to show methylation difference across genome for each tissue type and status that is biologically interpretable

```{r,eval=T}
memory.limit(size=10000)
grs.breast.tum <- gen.rat.set[,which(gen.rat.set$Tissue=="breast" & gen.rat.set$Status=="cancer")]
grs.breast.norm <- gen.rat.set[,which(gen.rat.set$Tissue=="breast" & gen.rat.set$Status=="normal")]
grs.colon.norm <- gen.rat.set[,which(gen.rat.set$Tissue=="colon" & gen.rat.set$Status=="normal")]
grs.lung.norm <- gen.rat.set[,which(gen.rat.set$Tissue=="lung" & gen.rat.set$Status=="normal")]

beta.breast.tum <- getBeta(grs.breast.tum)
beta.breast.norm <- getBeta(grs.breast.norm)
beta.colon.norm <- getBeta(grs.colon.norm)
beta.lung.norm <- getBeta(grs.lung.norm)

rm(grs.breast.tum,grs.breast.norm,grs.colon.norm,grs.lung.norm)

cpg.gr <- granges(gen.meth.set)
cpg.gr$beta.breast.tum <- rowMeans(beta.breast.tum)
cpg.gr$beta.breast.norm <- rowMeans(beta.breast.norm)
cpg.gr$beta.colon.norm <- rowMeans(beta.colon.norm)
cpg.gr$beta.lung.norm <- rowMeans(beta.lung.norm)

rm(beta.breast.tum,beta.breast.norm,beta.colon.norm,beta.lung.norm)

beta.status<- mgr$addDevice(cpg.gr,"Percent Methylation",type="bp",columns=c("beta.breast.tum","beta.breast.norm"))
```

Add gene expression barcode heatmap to top of workspace
-click on Vizualizations button
-click on "Add New Heatmap"
-Select "gexp_barcode" for the Gene Expression Barcode data and click "Next"
-Select breast tumor, breast normal, colon normal, and lung normal from the list (using Ctrl) and click "Finish"
-This shows the differences in gene expression between the different tissue types and statuses

Add MA plot of breast tissue methylation to vizualize which methylation difference points are important
-Use beta.breast.tum and beta.breast.norm that were already added to the Epiviz workspace 
-Use "Computed Measurements" to create plot with existing data

### 5. DMR Identificaiton:

Use bumphunter to find small differentially methylated regions (small deviations in methylation of CpG islands at the gene promoter scale, 1-2kb) between normal and tumor breast tissue

Bumphunting uses a cutoff value to determine the severity in the difference in methylation and determines if it is significant enough to label it as a differentially methylated region (DMR).

```{r,eval=F}
gen.rat.set.breast <- gen.rat.set[,which(gen.rat.set$Tissue=="breast")]
status <- pData(gen.rat.set.breast)$Status
X <- model.matrix(~status)
gr <- granges(gen.rat.set.breast)
chr <- as.factor(seqnames(gr))
pos <- start(gr)
cl <- clusterMaker(chr,pos,maxGap=500)
bumps <- bumphunter(gen.rat.set.breast,X,cluster=cl,cutoff=0.1,B=0)
```

Add hypo-methylation and hyper-methylation block tracks between normal and tumor breast tissue types

```{r, eval=F}
dmr.gr <- with(bumps$table,GRanges(chr,IRanges(start,end),area=area,value=value))
dmr.gr$type <- ifelse(abs(dmr.gr$value)<0.2, "neither", ifelse(dmr.gr$value<0,"hypo","hyper"))
table(dmr.gr$type)

hypo.dmr <- mgr$addDevice(dmr.gr[which(dmr.gr$type=="hypo")], "Hypo-methylated Regions")
hyper.dmr <- mgr$addDevice(dmr.gr[which(dmr.gr$type=="hyper")], "Hyper-methylated Regions")
```

Navigate the Epiviz workspace window to display the TRIM29 gene (located on chromosome 11 at base pairs 119513394-119513711)

```{r,eval=T}
mgr$navigate("chr11",119870000,120100000)
```

Look at transcription regulators, where differences in methylation have been found (ALX4, GATA5, MGMT, NEUROG1, SOX10, SREBP1, ST18, TRIM29,TP73)

These genes all have roles in tissue differentiation, cancer, or epigenetic regulation

Slideshow regions for genes of interest

ALX4.........chr11.....44282278.....44331716
GATA5........chr20.....61038553.....61051026
MGMT.........chr10.....131265454....131565783
NEUROG1......chr5......134869972....134871639
SOX10........chr22.....38368319.....38380539
SREBP1.......chr17.....15810399.....19837017
ST18.........chr8......53023392.....53130461
TP73.........chr1......3569129......3652765

```{r,eval=T}
IR <- IRanges(start=c(44282278,61038553,131265454,134869972,38368319,15810399,53023392,3569129),end=c(44331716,61051026,131565783,134871639,38380539,19837017,53130461,3652765))
IR <- IR*0.5
GR <- GRanges(c("chr11","chr20","chr10","chr5","chr22","chr17","chr8","chr1"),IR)
mgr$slideshow(GR)
=======
There is a CpG island at the transcription start site of TRIM29 that is differentially methylated 

```{r,eval=F}
library(doParallel)
cores <- detectCores()
registerDoParallel(cores)
status <- pData(brca.grs)$Status
#index <- which(seqnames(gen.rat.set)=="chr9")
#meth.set.9 <- meth.set[index,]
dmp <- dmpFinder(meth.set,type="categorical")
head(dmp)
```

#### B. Epiviz Workspace:

View difference in methylation amount between two different phenotypes 

In Epiviz, create a blocks track that shows the location of the DMPs and a block the shows the location of CpGs

```{r, eval=F}
cpg.gr <- granges(gen.meth.set)

cpgs <- mgr$addDevice(cpg.gr, "CpG Locations")
```

```{r, eval=F}
dmp.gr <- granges(gen.meth.set[rownames(dmp)])

dmps <- mgr$addDevice(dmp.gr, "Differentially Methylated Positions")
```

Add track of p-values to show which DMPs are significant (not significant if p > 0.05:

```{r, eval=F}
pvals <- mgr$addDevice(-log10(dmp$pval), "P-Values (-log10) of Differentially Methylated Positions")
getM(gen.rat.set)
```

Add track of smooth difference in methylation

```{r, eval=F}
beta <- getBeta(gen.rat.set)
beta.gr <- granges(gen.meth.set[rownames(beta)])
cpg.gr <- granges(gen.meth.set[rownames(bumps$fitted)])
meth.diff <- mgr$addDevice(cpg.gr,"Methylation Difference",type="bp")
``````

### 5. DMR Identificaiton:

#### A. R Analysis:

Use bumphunter to find small differentially methylated regions (small deviations in methylation of CpG islands at the gene promoter scale, 1-2kb)

Bumphunting uses a cutoff value to determine the severity in the difference in methylation and determines if it is significant enough to label it as a differentially methylated region (DMR).

Compare:
breast and colon
breast and lung 
breast and endometrium

Combine results to show common DMRs (about 120 loci) between normal breast tissue and the other three normal tissues

```{r,eval=F}
X <- model.matrix(~status)
gr <- granges(gen.meth.set)
chr=as.factor(seqnames(gr))
pos=start(gr)
cl=clusterMaker(chr,pos,maxGap=500)
bumps <- bumphunter(gen.rat.set,X,cluster=cl,cutoff=0.1,B=0)
```

#### B. Epiviz Workspace:

Look at transcription regulators, where differences in methylation have been found (ALX4, GATA5, MGMT, NEUROG1, SOX10, SREBP1, ST18, TRIM29,TP73)

Methylation differences for these genes between normal and cancer tissues 

These genes all have roles in tissue differentiation, cancer, or epigenetic regulation

Create a blocks track in Epiviz and add the DMR blocks to it (both hypo-methylated and hyper-methylated blocks).  Then, zoom in and find the region defined by the start and end values in the above table.  This would verify that there is indeed a small region of differentially methylated DNA specified by this region. 

```{r, eval=F}
dmr.gr <- with(bumps$table,GRanges(chr,IRanges(start,end),area=area,value=value))
dmr.gr$type <- ifelse(abs(dmr.gr$value)<0.2, "neither", ifelse(dmr.gr$value<0,"hypo","hyper"))
table(dmr.gr$type)

hypo.dmr <- mgr$addDevice(subset(dmr.gr,dmr.gr$type=="hypo"), "Hypo-methylated Regions")
hyper.dmr <- mgr$addDevice(subset(dmr.gr,dmr.gr$type=="hyper"), "Hyper-methylated Regions")
>>>>>>> origin/master
```

### 6. Block Identificaiton:

<<<<<<< HEAD
Load new Epiviz workspace to look at epigenetic differences between breast normal, colon normal, and lung normal tissues

```{r,eval=T}
mgr <- startEpiviz(workspace="Ago25FdPLVb",tryPorts=TRUE)
beta.tissue <- mgr$addDevice(cpg.gr,"Percent Methylation",type="bp",columns=c("beta.breast.norm","beta.colon.norm","beta.lung.norm"))
```
=======
#### A. R Analysis:
>>>>>>> origin/master

Identify large differentially methylated regions using blockFinder

In comparison, block finding is used to find differences in methylation on a much larger scale than bump hunting.

Block finding uses the cpgCollapse function to create clusters of neighboring open sea loci, grouping them into larger regions than bump hunting would, and then uses bump hunting on the averages of these regions to detect large differentially methylated regions.

```{r,eval=F}
<<<<<<< HEAD
memory.limit(size=10000)
gen.meth.set.breast <- gen.meth.set[,which(gen.meth.set$Tissue=="breast")]
#gen.meth.set.colon <- gen.meth.set[,which(gen.meth.set$Tissue=="colon")] 
#gen.meth.set.lung <- gen.meth.set[,which(gen.meth.set$Tissue=="lung")]
#gen.meth.set.norm <- gen.meth.set[,which(gen.meth.set$Status=="normal")]
status <- pData(gen.meth.set.breast)$Status
#tissue <- pData(gen.meth.set.norm)$Tissue
X <- model.matrix(~status)
#X <- model.matrix(~tissue)
cl <- cpgCollapse(gen.meth.set.breast)
#cl <- cpgCollapse(gen.meth.set.colon)
#cl <- cpgCollapse(gen.meth.set.lung)
#cl <- cpgCollapse(gen.meth.set.norm)
blocks <- blockFinder(cl$object,X,cluster=cl$blockInfo$pns,cutoff = 0.1)
```

Create a blocks track in Epiviz and add the hypo-methylation blocks to it. 
=======
#index <- which(seqnames(gen.rat.set)=="chr9")
#gen.meth.set.9 <- gen.meth.set[index,]
cl <- cpgCollapse(gen.meth.set)
blocks <- blockFinder(cl$object,X,cluster=cl$blockInfo$pns,cutoff = 0.1)
head(blocks$table)
```

#### B. Epiviz Workspace:

This can be shown by creating the hypo-methylation blocks in Epiviz.

Create a blocks track in Epiviz and add the hypo-methylation blocks to it.  Then, zoom in and find the region defined by the start and end values in the above table.  This would verify that there is indeed a large region of differentially methylated DNA specified by this region. The start and end points may not be exact depending on the resolution of the array.
>>>>>>> origin/master

```{r, eval=F}
seqnames <- blocks$table$chr
ranges <- IRanges(start=blocks$table$start,end=blocks$table$end)
blocks.gr <- GRanges(seqnames=seqnames,ranges=ranges)
<<<<<<< HEAD
meth.blocks <- mgr$addDevice(blocks.gr, "Differentially Methylated Blocks")
```

Add Epiviz slideshow to show top five differentially methylated promoter regionsAdd breast tissue differentially methylated regions to Epiviz workspace

```{r,eval=F}
top.blocks <- head(blocks$table,5)
top.blocks <- data.frame(top.blocks)
IR <- IRanges(start=top.blocks$start,end=top.blocks$end)
IR <- IR*0.5
GR <- GRanges(IR,seqnames=top.blocks$chr)
mgr$slideshow(GR)
```

Epiviz Workspace Screenshot

<img src="/Users/mwalter/Documents/screenshots/EpivizScreenshot.png" width="2000" height="1200" /> 
=======
mgr$service()
meth.blocks <- mgr$addDevice(blocks.gr, "Differentially Methylated Blocks")
mgr$service()
```

### 7. Useful Epiviz Workspace Additions:

Look more in depth, specifically at the TRIM29 gene (located on chromosome 11 at base pairs 119513394-119513711

Epiviz slideshow to show top five differentially methylated promoter regions

```{r, eval=F}
foldChange=mat[,"cancer"]-mat[,"normal"]
ind=order(foldChange,decreasing=TRUE)

# bounding 1Mb around each exon
slideshowRegions <- trim(rowRanges(sumexp)[ind] + 1000000L)
mgr$slideshow(slideshowRegions, n=5)
```

Use UCSC expression data

Transcription start site known to have differences in methylation (50% methylated in normal breast tissue, 100% in other normal tissues)
		-at the TRIM29 gene, methylation increases going from normal to tumor breast tissue, but decreases going from normal to other tumor types
		-look at methylation between different breast cancer types using UCSC expression data
		    -ER- basal normal-like (high expression of TRIM29)
		    -ER+ luminal tumor-like (low expression of TRIM29)
		    -inverse relationship with methyaltion levels at this regionStart Epiviz in web

Create MA plot to show average expression and difference in expression between breast normal and breast tumor at determined genes of interest

```{r, eval=F}
colData <- DataFrame(name=c("M","A"))
rownames(colData) <- colData$name

rowData <- gr
rowData$cpg <- names(gr)

fit <- lmFit(gen.rat.set,X)
cpg.sum.exp <- SummarizedExperiment(rowData=rowData,assays=SimpleList(ma=cbind(fit$coef[,2],fit$Amean)),colData=colData)
```
>>>>>>> origin/master
