---
title: Minfi Methylation Analysis with Epiviz Integration
author: Morgan Walter
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=TRUE)
```
 
## Introduction:

#### Purpose: 
To analyze Illumina 450k Methylation Array data to find differences in methylation at the TRIM29 gene located on chromosome 11 and other transcription regulators between various tumor samples (breast, lung, colon) and between normal and tumor breast cancer samples. 

#### How:  
Using the minfi and epivizr packages in Bioconductor and Epiviz as an interactive genomic tool, this protocol aims to walk through the analysis of the Illumina HumanMethylation450 Array data from the Cancer Genome Atlas (TCGA) project.

#### Why:
The TRIM29 gene is of interest because of its opposing functions in various cancers.  In some, it acts as a tumor suppressor (in breast cancers), and in others, it acts as an oncogene (lung cancers).  

## 1. Getting Started

Packages needed: Bioconductor, minfi, epivizr, GenomicFeatures (within Bioconductor)

```{r,eval=F}
source("http://bioconductor.org/biocLite.R")
biocLite("minfi")
biocLite("epivizr")
biocLite("GenomicFeatures")
biocLite("BSgenome.Hsapiens.UCSC.hg19")
biocLite("TxDb.Hsapiens.UCSC.hg19.knownGene")
```

## 2. Gather Data

In order to analyze the data, it must first be loaded into R.

Download TCGA 450k methylation array using the data matrix in the data portal

https://tcga-data.nci.nih.gov/tcga/dataAccessMatrix.htm

When selecting data, choose samples that end in 01 (TCGA-..-....-01) for solid tissue tumor samples and samples that end in 11 (TCGA-..-....-11) for solid tissue normal samples.

For this protocol, only breast tumor and breast normal tissue (from BRCA), colon normal (from COAD), lung normal (from LUSC) is loaded.

Follow along with the screenshots:

BRCA download settings
![BRCA download settings](images/brca_mat.png)

BRCA-tumor chosen samples
![BRCA-tumor chosen samples](images/brca_tum_dat.png)  

BRCA-normal chosen samples
![BRCA-normal chosen samples ](images/brca_norm_dat.png)

COAD download settings
![COAD download settings](images/coad_mat.png)

COAD-normal chosen samples
![COAD-normal chosen samples](images/coad_norm_dat.png)

LUSC download settings
![LUSC download settings](images/lusc_mat.png)  

LUSC-normal chosen samples
![LUSC-normal chosen samples](images/lusc_norm_dat.png)  

For each cancer type, once the correct settings are selected, click the "Build Archive" button and then download the data.

After unzipping the files that are emailed to you, save each cancer type (BRCA, COAD, LUSC) as different folders in "datadir", like shown below.

Load breast tumor and breast normal data

```{r,eval=F}
library(minfi)
datadir <- "/Users/mwalter/Documents/methylation_files/breast"

clinicalDir <- file.path(datadir,"Clinical/Biotab")
sample_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_sample_brca.txt"),sep="\t",stringsAsFactors=FALSE)
keep <- sample_tab$sample_type %in% c("Primary Tumor", "Solid Tissue Normal")
sample_tab <- sample_tab[keep,]

patient_id <- unique(sapply(strsplit(sample_tab$bcr_sample_barcode,split="-"), function(x) paste(x[1:3],collapse="-")))

tumor_sample_id <- sample_tab$bcr_sample_uuid[sample_tab$sample_type=="Primary Tumor"]
normal_sample_id <- sample_tab$bcr_sample_uuid[sample_tab$sample_type=="Solid Tissue Normal"]

# read tumor data
tumor_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_tumor_sample_brca.txt"),sep="\t",stringsAsFactors=FALSE)

tab <- merge(sample_tab, tumor_tab, by="bcr_sample_uuid", suffixes=c(".sample",".tumor"),all.x=TRUE)

# read normal data
normal_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_normal_control_brca.txt"),sep="\t",stringsAsFactors=FALSE)
tab <- merge(tab, normal_tab, by="bcr_sample_uuid", suffixes=c(".tumor",".normal"),all.x=TRUE)

tab$bcr_patient_barcode <- tab$bcr_patient_barcode.tumor

ii <- is.na(tab$bcr_patient_barcode)
tab$bcr_patient_barcode[ii] <- tab$bcr_patient_barcode.normal[ii]

# read patient data
patient_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_clinical_patient_brca.txt"),sep="\t",stringsAsFactors=FALSE)
names(patient_tab) <- paste("patient",names(patient_tab),sep=".")
tmp <- merge(tab,patient_tab,by.x="bcr_patient_barcode",by.y="patient.bcr_patient_barcode",all.x=TRUE,suffixes=c(".sample",".patient"))
tab <- tmp

# read meth metadata
methMetaDir <- file.path(datadir,"METADATA/JHU_USC__HumanMethylation450")
methMeta_tab <- read.delim(file.path(methMetaDir,"jhu-usc.edu_BRCA.HumanMethylation450.1.9.0.sdrf.txt"),sep="\t",stringsAsFactors=FALSE)

sample_barcode <- sapply(strsplit(methMeta_tab$Comment..TCGA.Barcode.,split="-"),function(x) paste(x[1:4],collapse="-"))
m <- match(tab$bcr_sample_barcode,sample_barcode)
tab$Basename <- gsub("_Grn\\.idat","",methMeta_tab$Array.Data.File[m])
tab <- tab[!is.na(tab$Basename),]

basedir <- file.path(datadir,"DNA_Methylation/JHU_USC__HumanMethylation450/Level_1")
tab$Basename <- file.path(basedir,tab$Basename)
keep <- file.exists(paste(tab$Basename,"_Grn.idat",sep=""))
breast_targets <- tab
objs <- grep("tab",ls(),value=TRUE)
rm(list=objs)
objs <- grep("dir",ls(),value=TRUE,ignore=TRUE)
rm(list=objs)

nms <- names(breast_targets)
targets.breast <- breast_targets[nms]

targets.breast$Status <- factor(ifelse(targets.breast$sample_type=="Primary Tumor","cancer","normal"),levels=c("normal","cancer"))
targets.breast$Tissue <- tolower(targets.breast$patient.tumor_tissue_site)
targets.breast$Sex <- targets.breast$patient.gender
```

Load colon normal data

```{r,eval=F}
datadir <- "/Users/mwalter/Documents/methylation_files/colon"

clinicalDir <- file.path(datadir,"Clinical/Biotab")
sample_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_sample_coad.txt"),sep="\t",stringsAsFactors=FALSE)
keep <- sample_tab$sample_type %in% c("Primary Tumor", "Solid Tissue Normal")
sample_tab <- sample_tab[keep,]

patient_id <- unique(sapply(strsplit(sample_tab$bcr_sample_barcode,split="-"), function(x) paste(x[1:3],collapse="-")))

tumor_sample_id <- sample_tab$bcr_sample_uuid[sample_tab$sample_type=="Primary Tumor"]
normal_sample_id <- sample_tab$bcr_sample_uuid[sample_tab$sample_type=="Solid Tissue Normal"]

# read tumor data
tumor_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_tumor_sample_coad.txt"),sep="\t",stringsAsFactors=FALSE)

tab <- merge(sample_tab, tumor_tab, by="bcr_sample_uuid", suffixes=c(".sample",".tumor"),all.x=TRUE)

# read normal data
normal_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_normal_control_coad.txt"),sep="\t",stringsAsFactors=FALSE)
tab <- merge(tab, normal_tab, by="bcr_sample_uuid", suffixes=c(".tumor",".normal"),all.x=TRUE)

tab$bcr_patient_barcode <- tab$bcr_patient_barcode.tumor

ii <- is.na(tab$bcr_patient_barcode)
tab$bcr_patient_barcode[ii] <- tab$bcr_patient_barcode.normal[ii]

# read patient data
patient_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_clinical_patient_coad.txt"),sep="\t",stringsAsFactors=FALSE)
names(patient_tab) <- paste("patient",names(patient_tab),sep=".")
tmp <- merge(tab,patient_tab,by.x="bcr_patient_barcode",by.y="patient.bcr_patient_barcode",all.x=TRUE,suffixes=c(".sample",".patient"))
tab <- tmp

# read meth metadata
methMetaDir <- file.path(datadir,"METADATA/JHU_USC__HumanMethylation450")
methMeta_tab <- read.delim(file.path(methMetaDir,"jhu-usc.edu_COAD.HumanMethylation450.1.9.0.sdrf.txt"),sep="\t",stringsAsFactors=FALSE)

sample_barcode <- sapply(strsplit(methMeta_tab$Comment..TCGA.Barcode.,split="-"),function(x) paste(x[1:4],collapse="-"))
m <- match(tab$bcr_sample_barcode,sample_barcode)
tab$Basename <- gsub("_Grn\\.idat","",methMeta_tab$Array.Data.File[m])
tab <- tab[!is.na(tab$Basename),]

basedir <- file.path(datadir,"DNA_Methylation/JHU_USC__HumanMethylation450/Level_1")
tab$Basename <- file.path(basedir,tab$Basename)
keep <- file.exists(paste(tab$Basename,"_Grn.idat",sep=""))
colon_targets <- tab
objs <- grep("tab",ls(),value=TRUE)
rm(list=objs)
objs <- grep("dir",ls(),value=TRUE,ignore=TRUE)
rm(list=objs)

nms <- names(colon_targets)
targets.colon <- colon_targets[nms]

targets.colon$Status <- factor(ifelse(targets.colon$sample_type=="Primary Tumor","cancer","normal"),levels=c("normal","cancer"))
targets.colon$Tissue <- tolower(targets.colon$patient.tumor_tissue_site)
targets.colon$Sex <- targets.colon$patient.gender
```

Load lung normal data

```{r,eval=F}
datadir <- "/Users/mwalter/Documents/methylation_files/lung"


clinicalDir <- file.path(datadir,"Clinical/Biotab")
sample_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_sample_lusc.txt"),sep="\t",stringsAsFactors=FALSE)
keep <- sample_tab$sample_type %in% c("Primary Tumor", "Solid Tissue Normal")
sample_tab <- sample_tab[keep,]

patient_id <- unique(sapply(strsplit(sample_tab$bcr_sample_barcode,split="-"), function(x) paste(x[1:3],collapse="-")))

tumor_sample_id <- sample_tab$bcr_sample_uuid[sample_tab$sample_type=="Primary Tumor"]
normal_sample_id <- sample_tab$bcr_sample_uuid[sample_tab$sample_type=="Solid Tissue Normal"]

# read tumor data
tumor_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_tumor_sample_lusc.txt"),sep="\t",stringsAsFactors=FALSE)

tab <- merge(sample_tab, tumor_tab, by="bcr_sample_uuid", suffixes=c(".sample",".tumor"),all.x=TRUE)

# read normal data
normal_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_biospecimen_normal_control_lusc.txt"),sep="\t",stringsAsFactors=FALSE)
tab <- merge(tab, normal_tab, by="bcr_sample_uuid", suffixes=c(".tumor",".normal"),all.x=TRUE)

tab$bcr_patient_barcode <- tab$bcr_patient_barcode.tumor

ii <- is.na(tab$bcr_patient_barcode)
tab$bcr_patient_barcode[ii] <- tab$bcr_patient_barcode.normal[ii]

# read patient data
patient_tab <- read.delim(file.path(clinicalDir,"nationwidechildrens.org_clinical_patient_lusc.txt"),sep="\t",stringsAsFactors=FALSE)
names(patient_tab) <- paste("patient",names(patient_tab),sep=".")
tmp <- merge(tab,patient_tab,by.x="bcr_patient_barcode",by.y="patient.bcr_patient_barcode",all.x=TRUE,suffixes=c(".sample",".patient"))
tab <- tmp

# read meth metadata
methMetaDir <- file.path(datadir,"METADATA/JHU_USC__HumanMethylation450")
methMeta_tab <- read.delim(file.path(methMetaDir,"jhu-usc.edu_LUSC.HumanMethylation450.1.7.0.sdrf.txt"),sep="\t",stringsAsFactors=FALSE)

sample_barcode <- sapply(strsplit(methMeta_tab$Comment..TCGA.Barcode.,split="-"),function(x) paste(x[1:4],collapse="-"))
m <- match(tab$bcr_sample_barcode,sample_barcode)
tab$Basename <- gsub("_Grn\\.idat","",methMeta_tab$Array.Data.File[m])
tab <- tab[!is.na(tab$Basename),]

basedir <- file.path(datadir,"DNA_Methylation/JHU_USC__HumanMethylation450/Level_1")
tab$Basename <- file.path(basedir,tab$Basename)
keep <- file.exists(paste(tab$Basename,"_Grn.idat",sep=""))
lung_targets <- tab
objs <- grep("tab",ls(),value=TRUE)
rm(list=objs)
objs <- grep("dir",ls(),value=TRUE,ignore=TRUE)
rm(list=objs)

nms <- names(lung_targets)
targets.lung <- lung_targets[nms]

targets.lung$Status <- factor(ifelse(targets.lung$sample_type=="Primary Tumor","cancer","normal"),levels=c("normal","cancer"))
targets.lung$Tissue <- tolower(targets.lung$patient.tumor_tissue_site)
targets.lung$Sex <- targets.lung$patient.gender

rm(list=ls()[!(ls() %in% c('targets.breast','targets.colon','targets.lung'))])
```

Merge and read data

```{r,eval=F}
merge <- merge(targets.breast,targets.colon,all=TRUE)
targets <- merge(merge,targets.lung,all=TRUE)
targets <- targets[which(file.exists(paste0(targets$Basename,"_Grn.idat"))),]

memory.limit(size=10000)
rg.set <- read.450k(targets$Basename,verbose=TRUE)
pData(rg.set) <- targets

table(targets$Tissue,targets$Status)
```

## 3. Preprocessing

Process data into usable objects for the minfi package.

Need preprocessing to modify data from IDAT files into useable data sets, like an RGChannelSet, MethylSet, GenomicMethylSet, RatioSet, or GenomicRatioSet for the minfi functions to work. -Most of the analysis will be done on the GenomicRatioSet datatype.


Use minfi's variety of preprocess...() functions to convert data sets into usable forms

How to preprocess into useful objects: 

Input                  | Processing Function    | Output           | Analytic Use
---------------------- | ---------------------- | ---------------- | --------------------------
Raw data (IDAT files)  | read.450k.exp()        | RGChannelSet     | output of reading data   
RGChannelSet           | preprocessIllumina()   | MethylSet        | dmpFinder method
MethylSet              | mapToGenome()          | GenomicMethylSet | blockFinder method
GenomicMethylSet       | ratioConvert()         | GenomicRatioSet  | bumphunter method

```{r,eval=F}
meth.set <- preprocessIllumina(rg.set)
gen.meth.set <- mapToGenome(meth.set)
gen.rat.set <- ratioConvert(gen.meth.set,type="Illumina")
```

## 4. DMP Vizualization

#### Load Epiviz Workspace #1:

Use Epiviz to vizualize the results found in R session

Start Epiviz in web, starting from a pre-existing workspace (delete all tracks but "Genes" and "CpG Islands")

```{r, eval=F}
require(epivizr)
mgr <- startEpiviz(workspace="Sp0EK3l2LUu",tryPorts=TRUE)
#mgr$service()
```

#### Analysis:

Use dmpFinder to find differentially methylated positions to find CpG methylation differences that correspond to a particular phenotype (in this case breast normal and breast tumor)

dmpFinder uses an F-test between two groups of different phenotypes to determine which CpG loci are significantly differentially methylated

Look at difference between normal breast and tumor breast tissues

Start with the genes track at the top

Add promoter regions in 

```{r,eval=F}
library(GenomicFeatures)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
promoter.regions <- promoters(TxDb.Hsapiens.UCSC.hg19.knownGene,upstream=2000,downstream=200)
promoters <- mgr$addDevice(promoter.regions, "Promoters")
```

In Epiviz, add a blocks track to show the location of the DMPs

```{r,eval=F}
library(doParallel)
cores <- detectCores()
registerDoParallel(cores)
meth.set.breast <- meth.set[,which(meth.set$Tissue=="breast")]
dmp <- dmpFinder(meth.set.breast,meth.set.breast$Status,type="categorical")
head(dmp)
dmp.gr <- granges(gen.meth.set[rownames(dmp)])
dmps <- mgr$addDevice(dmp.gr, "Differentially Methylated Positions")
```

Add track of p-values underneath to show which DMPs are significant (not significant if p > 0.05):

```{r, eval=F}
dmp.gr$pval <- -log10(dmp$pval)
pvals <- mgr$addDevice(dmp.gr, "P-Values (-log10) of DMPs",type="bp",columns="pval")
```

Add beta-values to show methylation difference across genome for each tissue type and status that is biologically interpretable (percent methylation)

```{r,eval=F}
memory.limit(size=10000)
grs.breast.tum <- gen.rat.set[,which(gen.rat.set$Tissue=="breast" & gen.rat.set$Status=="cancer")]
grs.breast.norm <- gen.rat.set[,which(gen.rat.set$Tissue=="breast" & gen.rat.set$Status=="normal")]
grs.colon.norm <- gen.rat.set[,which(gen.rat.set$Tissue=="colon" & gen.rat.set$Status=="normal")]
grs.lung.norm <- gen.rat.set[,which(gen.rat.set$Tissue=="lung" & gen.rat.set$Status=="normal")]

beta.breast.tum <- getBeta(grs.breast.tum)
beta.breast.norm <- getBeta(grs.breast.norm)
beta.colon.norm <- getBeta(grs.colon.norm)
beta.lung.norm <- getBeta(grs.lung.norm)

rm(grs.breast.tum,grs.breast.norm,grs.colon.norm,grs.lung.norm)

cpg.gr <- granges(gen.meth.set)
cpg.gr$beta.breast.tum <- rowMeans(beta.breast.tum)
cpg.gr$beta.breast.norm <- rowMeans(beta.breast.norm)
cpg.gr$beta.colon.norm <- rowMeans(beta.colon.norm)
cpg.gr$beta.lung.norm <- rowMeans(beta.lung.norm)

rm(beta.breast.tum,beta.breast.norm,beta.colon.norm,beta.lung.norm)

beta.tissue <- mgr$addDevice(cpg.gr,"Percent Methylation",type="bp",columns=c("beta.breast.tum","beta.breast.norm","beta.colon.norm","beta.lung.norm"))
```

Add gene expression barcode heatmap to top of workspace
-click on Vizualizations button
-click on "Add New Heatmap"
-Select "gexp_barcode" for the Gene Expression Barcode data and click "Next"
-Select breast tumor, breast normal, colon normal, and lung normal from the list (using Ctrl) and click "Finish"
-This shows the differences in gene expression between the different tissue types and statuses

Add an MA plot of breast tissue methylation to vizualize which methylation difference points are important
-Use beta.breast.tum and beta.breast.norm that were already added to the Epiviz workspace 
-Use "Computed Measurements" to create plot with existing data

Navigate the Epiviz workspace window to display the TRIM29 gene (located on chromosome 11 at base pairs 119513394-119513711)

```{r,eval=F}
mgr$navigate("chr11",119870000,120100000)
```

The ending workspace should look like this: 

DMP Vizualization
![DMP Vizualization](images/dmp_viz.png)  

## 5. DMR Vizualization

#### Load Epiviz Workspace #2:

Load new Epiviz workspace to look at epigenetic differences between breast normal, colon normal, and lung normal tissues (delete all tracks except for "Genes" and "CpG Islands")

```{r,eval=F}
require(epivizr)
mgr <- startEpiviz(workspace="y0bTBoRr2Bl",tryPorts=TRUE)
```

#### Analysis:

Add promoter regions in 

```{r,eval=F}
library(GenomicFeatures)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
promoter.regions <- promoters(TxDb.Hsapiens.UCSC.hg19.knownGene,upstream=2000,downstream=200)
promoters <- mgr$addDevice(promoter.regions, "Promoters")
```

Add smooth beta value track showing differences in methylation between the three tissues

```{r,eval=F}
beta.tissue <- mgr$addDevice(cpg.gr,"Percent Methylation",type="bp",columns=c("beta.breast.tum","beta.breast.norm","beta.colon.norm","beta.lung.norm"))
```

Use bumphunter to find small differentially methylated regions (small deviations in methylation of CpG islands at the gene promoter scale, 1-2kb) between normal and tumor breast tissue

Bumphunting uses a cutoff value to determine the severity in the difference in methylation and determines if it is significant enough to label it as a differentially methylated region (DMR).

```{r,eval=F}
gen.rat.set.breast <- gen.rat.set[,which(gen.rat.set$Tissue=="breast")]
status <- pData(gen.rat.set.breast)$Status
X <- model.matrix(~status)
gr <- granges(gen.rat.set.breast)
chr <- as.factor(seqnames(gr))
pos <- start(gr)
cl <- clusterMaker(chr,pos,maxGap=500)
bumps <- bumphunter(gen.rat.set.breast,X,cluster=cl,cutoff=0.1,B=0)
```

Add general differentially methylated regions, hypo-methylated regions, and hyper-methylated regions block tracks between normal and tumor breast tissue types

The hypo-methylated regions and the hyper-methylated regions should add up to make the total differentially methylated regions

```{r, eval=F}
dmr.gr <- with(bumps$table,GRanges(chr,IRanges(start,end),area=area,value=value))
dmr.gr$type <- ifelse(abs(dmr.gr$value)<0.1, "neither", ifelse(dmr.gr$value<0,"hypo","hyper"))
table(dmr.gr$type)

bumps.dmr <- mgr$addDevice(dmr.gr, "Breast DMRs")
hypo.dmr <- mgr$addDevice(dmr.gr[which(dmr.gr$type=="hypo")], "Hypo-methylated Regions")
hyper.dmr <- mgr$addDevice(dmr.gr[which(dmr.gr$type=="hyper")], "Hyper-methylated Regions")
```

Methylation difference
```{r,eval=F}
dmr.value <- mgr$addDevice(dmr.gr,"Methylation Difference",type="bp",columns=c("value"))
```

Display differentially methylated promoters by finding overlaps in promoters and breast DMRs

```{r,eval=F}
dmr.promoters <- subsetByOverlaps(dmr.gr,promoter.regions)
diff.meth.promoters <- mgr$addDevice(dmr.promoters, "Differentially Methylated Promoters")
```

Add gene expression barcode heatmap to top of workspace (same as in DMP Vizualization section)
-click on Vizualizations button
-click on "Add New Heatmap"
-Select "gexp_barcode" for the Gene Expression Barcode data and click "Next"
-Select breast tumor, breast normal, colon normal, and lung normal from the list (using Ctrl) and click "Finish"
-This shows the differences in gene expression between the different tissue types and statuses 

Add an MA plot of breast tissue expression to vizualize differences in gene expression that are important.  This MA plot is different than the one used in the DMP Vizualization section as it shows the differences in gene expression rather than DNA methylation. 
-Use "Computed Measurements" with breast tumor and breast normal measurements under "gexp_barcode" datasource group to create plot with existing data

Slideshow regions for top five differentially methylated promoter regions

```{r,eval=F}
hypo.promoters <- dmr.promoters[which(dmr.promoters$type=="hypo")]
top.promoters <- head(hypo.promoters,5)
start(top.promoters) <- start(top.promoters) - 100000
end(top.promoters) <- end(top.promoters) + 100000
mgr$slideshow(top.promoters)
```

Look at transcription regulators, where differences in methylation have been found (ALX4, GATA5, MGMT, NEUROG1, SOX10, SREBP1, ST18, TRIM29,TP73)

These genes all have roles in tissue differentiation, cancer, or epigenetic regulation

Slideshow regions for genes of interest

ALX4.........chr11.....44282278.....44331716
GATA5........chr20.....61038553.....61051026
MGMT.........chr10.....131265454....131565783
NEUROG1......chr5......134869972....134871639
SOX10........chr22.....38368319.....38380539
SREBP1.......chr17.....15810399.....19837017
ST18.........chr8......53023392.....53130461
TP73.........chr1......3569129......3652765

```{r,eval=F}
IR <- IRanges(start=c(44282278,61038553,131265454,134869972,38368319,15810399,53023392,3569129),end=c(44331716,61051026,131565783,134871639,38380539,19837017,53130461,3652765))
IR <- IR*0.5
GR <- GRanges(c("chr11","chr20","chr10","chr5","chr22","chr17","chr8","chr1"),IR)
mgr$slideshow(GR)
```

The ending workspace should look like this: 

DMR Vizualization
![DMR Vizualization](images/dmr_viz.png)  

## 6. Block Vizualization

#### Load Epiviz Workspace #3:

Load new Epiviz workspace to look at epigenetic differences between breast normal, colon normal, and lung normal tissues

```{r,eval=F}
require(epivizr)
mgr <- startEpiviz(workspace="GPPsmSNaEHu",tryPorts=TRUE)
```

#### Analysis:

Add promoter regions in 

```{r,eval=F}
library(GenomicFeatures)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
promoter.regions <- promoters(TxDb.Hsapiens.UCSC.hg19.knownGene,upstream=2000,downstream=200)
promoters <- mgr$addDevice(promoter.regions, "Promoters")
```

Add smooth beta value track showing differences in methylation between the three tissues

```{r,eval=F}
beta.tissue <- mgr$addDevice(cpg.gr,"Percent Methylation",type="bp",columns=c("beta.breast.tum","beta.breast.norm","beta.colon.norm","beta.lung.norm"))
```

Breast Beta Methylation Difference

```{r,eval=FALSE}
cpg.gr$beta.breast.diff <- cpg.gr$beta.breast.tum-cpg.gr$beta.breast.norm
meth.diff <- mgr$addDevice(cpg.gr,"Methylation Beta Difference",type="bp",columns="beta.breast.diff")
```

Identify large differentially methylated regions using blockFinder

In comparison, block finding is used to find differences in methylation on a much larger scale than bump hunting.

Block finding uses the cpgCollapse function to create clusters of neighboring open sea loci, grouping them into larger regions than bump hunting would, and then uses bump hunting on the averages of these regions to detect large differentially methylated regions.

```{r,eval=F}
memory.limit(size=10000)
gen.meth.set.breast <- gen.meth.set[,which(gen.meth.set$Tissue=="breast")]
status <- pData(gen.meth.set.breast)$Status
X <- model.matrix(~status)
cl <- cpgCollapse(gen.meth.set.breast)
blocks <- blockFinder(cl$object,X,cluster=cl$blockInfo$pns,cutoff = 0.1)
```

Add general blocks, hypo-methylated blocks, and hyper-methylated blocks block tracks between normal and tumor breast tissue types

The hypo-methylated blocks and the hyper-methylated blocks should add up to make the total differentially methylated blocks

```{r, eval=F}
blocks.gr <- with(blocks$table,GRanges(chr,IRanges(start,end),area=area,value=value))
blocks.gr$type <- ifelse(abs(blocks.gr$value)<0.1, "neither", ifelse(blocks.gr$value<0,"hypo","hyper"))
table(blocks.gr$type)

meth.blocks <- mgr$addDevice(blocks.gr, "Breast Blocks")
hypo.blocks <- mgr$addDevice(blocks.gr[which(blocks.gr$type=="hypo")],"Hypo-methylated Blocks")
hyper.blocks <- mgr$addDevice(blocks.gr[which(blocks.gr$type=="hyper")],"Hyper-methylated Blocks")
```

Display differentially methylated genes by finding overlaps in promoters and breast DMRs

```{r,eval=F}
library(GenomicFeatures)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
gene.regions <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene,columns="gene_id",single.strand.genes.only=TRUE)
block.genes <- subsetByOverlaps(blocks.gr,gene.regions)
diff.meth.genes <- mgr$addDevice(block.genes,"Differentially Methylated Genes")
```

Add Epiviz slideshow to show top five differentially methylated gene regions

```{r,eval=F}
top.genes <- head(block.genes,5)
start(top.genes) <- start(top.genes) - 1000000
end(top.genes) <- end(top.genes) + 1000000
mgr$slideshow(top.genes)
```

The ending workspace should look like this: 

Block Vizualization
![Block Vizualization](images/block_viz.png) 
